### PR TITLE
RFC: Quantizers as subclasses of keras.layer.Layer

### LINKED ISSUE #246: Documentation (Regarding to JOSS review)
Some suggestions:
In the quantizers documentation:
1. It seems that SteSign/ste_sign (if they are the same) is repeated twice [here](https://larq.dev/api/quantizers/#ste_sign). The same for swish_sign/SwishSign.
2. These quantizers (magnitude_aware_sign,swish_sign)have no formula, nor description.

### PR DESCRIPTION
**Pull request by lgeiger**
This PR implements Proposal 1 and 2 of #343. The diff is quite large so it makes sense to carefully review this commit by commit.

Closes #246

### CODE REVIEW

File: larq/quantizers.py [hunk 1]
```diff
@@ -153,70 +83,8 @@ def _scaled_sign(x):  # pragma: no cover
     return 1.3 * ste_sign(x)
 
 
-@utils.set_precision(1)
-def magnitude_aware_sign(x, clip_value=1.0):
```
[the remark this answers is not in the document]
Comment (jneeven) [reply larq/quantizers.py]: I see that you eventually removed all of them, so I guess it doesn't matter

**Comment by jneeven** [comment]
I think this is overall a very nice change, perhaps it would be nice to have a `QuantizerLayer` superclass that we can add metrics to?

**Comment by lgeiger** [comment]
> Whats up with ApproxSign? Was it not in here before? (referring to your first commit)

Yes I think it wasn't here because it doesn't have any configurable parameters like `clip_value` (or maybe we just forgot).

> I think this is overall a very nice change, perhaps it would be nice to have a `QuantizerLayer` superclass that we can add metrics to?

Yes I think that would be the right thing to do. I am not sure yet if it should be a public API users could use to write their own quantizers, of if it should just be an implementation detail. I still need to check why tests are failing for TF != 2.1.0 before I'll start cleaning this PR up.

File: docs/guides/key-concepts.md [hunk 2]
```diff
@@ -3,7 +3,7 @@ Below, we summarize the key concepts you need to understand to work with BNNs.
 
 ## Quantizer
 
-The transformation from high-precision Neural Networks to Quantized Neural Networks (QNNs) is achieved by [quantization](https://en.wikipedia.org/wiki/Quantization_(signal_processing)).
+The transformation from high-precision Neural Networks to Quantized Neural Networks (QNNs) is achieved by [quantization](<https://en.wikipedia.org/wiki/Quantization_(signal_processing)>).
```
[the remark this answers is not in the document]
Comment (lgeiger) [reply docs/guides/key-concepts.md]: :+1: I think that was [Prettier](https://prettier.io/) just being confused.

**Comment by lgeiger** [comment]
This should now be ready for a thorough review

File: docs/guides/bnn-architecture.md [hunk 3]
```diff
@@ -19,8 +19,8 @@ import larq as lq
 ...
 x_out = lq.layers.QuantDense(
     units=1024,
-    input_quantizer=lq.quantizers.ste_sign,
-    kernel_quantizer=lq.quantizers.ste_sign,
+    input_quantizer=lq.quantizers.SteSign(clip_value=1.0),
+    kernel_quantizer=lq.quantizers.SteSign(clip_value=1.0),
     kernel_constraint=lq.constraints.WeightClip(clip_value=1.0),
```
Comment (leonoverweel) [inline docs/guides/bnn-architecture.md]: It's nice that the `SteSign()` instantiations are now more similar to the `WeightClip()` ones.

File: larq/conftest.py [hunk 4]
```diff
@@ -19,6 +19,7 @@ def graph_mode():
     with context.graph_mode():
         with tf.compat.v1.Session().as_default():
             yield
+            tf.keras.backend.clear_session()
```
Comment (leonoverweel) [inline larq/conftest.py]: Why is this necessary?

File: larq/quantized_variable_test.py [hunk 5]
```diff
@@ -21,7 +21,7 @@ def test_inheritance(distribute_scope):
     assert isinstance(quantized_variable, DistributedVariable) is distribute_scope  # type: ignore
 
 
-def test_read(distribute_scope, eager_and_graph_mode):
+def test_read(eager_and_graph_mode, distribute_scope):
```
Comment (leonoverweel) [inline larq/quantized_variable_test.py]: Why are these now all flipped?

File: larq/conftest.py (hunk 4 above)
Comment (lgeiger) [reply larq/conftest.py]: This is required for TensorFlow < 2.1 see f43aa7b25495a6ede2f9139f10559dfd8e450713

File: larq/quantized_variable_test.py (hunk 5 above)
Comment (lgeiger) [reply larq/quantized_variable_test.py]: Due to the change above, we need to set the `graph_mode` fixture before setting a distribution scope, otherwise the cleanup will error due to the graph beeing clean up, before exiting the distribution scope.
